### ISSUE TITLE
NullAway doesn't recognize Guava Preconditions checks

### ISSUE THREAD

**Issue by MatteoJoliveau** [uber/NullAway#47]
I'm using Google Guava [Preconditions](https://google.github.io/guava/releases/23.0/api/docs/com/google/common/base/Preconditions.html) to raise errors on null parameters. So basically, when I write something like:  
```java
Preconditions.checkArgument(item != null, "Unable to find an element with name '%s'.", name);
return item;
```
`item` is null safe, but NullAway still raises an error.

![image](https://user-images.githubusercontent.com/9077612/32056349-b8ae8a6a-ba64-11e7-9029-6b522346f238.png)

**Comment by msridhar** [comment]
Thanks for the report! I think we should add default support for `Preconditions.checkNotNull`; we have this supported internally. `checkArgument` is trickier since the API does more than nullness checking. @lazaroclapp could you look into this?

**Comment by lazaroclapp** [comment]
I believe `com.google.common.base.Preconditions.checkNotNull` is already supported in open source NullAway too. @MatteoJoliveau If you can use that method, then that's the fastest workaround right now. If this does not work, please let me know.

Supporting `Preconditions.checkArgument` can certainly be done with a handler, similar to how we propagate nullness information from stream `.filter()` operations. I will take a look at this when I can. It is worth thinking about the general case of adding library models where we know execution will terminate if a certain argument evaluates to either `true` or `false`, and then treat this method as an special case. As always, handlers come with a performance cost, however. Are there any `checkArgument` calls which check for nullness and which would be harder to express as `checkNotNull`?

**Comment by msridhar** [comment]
Ah, I didn't realize we already support `checkNotNull`.  We can think about supporting `checkArgument` in the future, but I think it will be low priority given that `checkNotNull` is available, due to performance concerns.

**Comment by MatteoJoliveau** [comment]
I will try to use `checkNotNull` since it's not that different from what I'm already doing (checking `item != null` is basically a check not null). 
But if you could support `checkArgument` too in the future it would be great, it gives a bit more freedom to the end users when using Preconditions.

**Comment by msridhar** [comment]
Yes, that is true.  If we can find a way to support `checkArgument` without too much performance penalty we will give it a shot.

**Comment by ben-manes** [comment]
`Objects.requireNonNull` should also be supported.

**Comment by msridhar** [comment]
@jrepe no concrete plans to prioritize this.  I would be open to accepting a PR if we could show it doesn't impact overhead too much (and I don't think it should).

**Comment by ketkarameya** [comment]
Another flavor of this is scenario is- 
`Preconditions.checkArgument(item != null && someOtherCondition(item), "Unable to find an element with name '%s'.", name);`
(Note I have a conjunction expression as the first argument).

**Comment by lazaroclapp** [comment]
This shows up enough internally and asked about externally often enough that we should probably prioritize it above `lowpriority`. Most cases seem to be reducible to `Preconditions.checkNotNull(...)`. But, as @ketkarameya 's example shows... some are a bit more involved.

There are also cases like this, where there is no `null` check, but there is a call that implies null checking:

```
Preconditions.checkArgument(Strings.isNullOrEmpty(s), "blah")
```

**Comment by lazaroclapp** [comment]
So, I see `AnalysisResult.getStoreBefore(...)`, but the problem is that we need to query this _while_ performing the analysis. Any way to get an `AnalysisResult` that I can find requires... well, running the analysis on the method first, which at best leaves us repeating work twice for each method.

I think a better alternative is to use handlers to move the APs around. The easy part of that is using `.onDataflowVisitMethodInvocation(...)` [here](https://github.com/uber/NullAway/blob/master/nullaway/src/main/java/com/uber/nullaway/dataflow/AccessPathNullnessPropagation.java#L895) to re-add them to the stores after the `Preconditions` call, getting it from some map maintained inside the handler. The hard part is finding all the types of `visitX` for expressions that can be passed to that `Preconditions.checkArgument(...)` call and saving the stores then...

**Comment by msridhar** [comment]
In the Checker Framework, the standard transfer function type `CFTransfer` keeps a reference to the `Analysis` object:

https://github.com/typetools/checker-framework/blob/95dd4483eeba563aa6755c9a4559a90db2e606a6/framework/src/main/java/org/checkerframework/framework/flow/CFTransfer.java

The `Analysis` type allows for getting values of `Node`s and `Tree`s while the analysis is running:

https://github.com/typetools/checker-framework/blob/95dd4483eeba563aa6755c9a4559a90db2e606a6/dataflow/src/main/java/org/checkerframework/dataflow/analysis/Analysis.java#L118-L136

I think the right thing to do here may be to re-jigger things so that `AccessPathNullnessPropagation` has a reference to the running `Analysis`.  This way we can get the current result for nodes without analyzing methods twice.

**Comment by lazaroclapp** [comment]
So, we already can get values for `Node`s and `Tree`s without that, just from the `TransferInput<Nullness, NullnessStore>` parameter [here](https://github.com/uber/NullAway/blob/master/nullaway/src/main/java/com/uber/nullaway/dataflow/AccessPathNullnessPropagation.java#L885). You can then do `input.getValueOfSubNode([node])`, which should be the same as `Analysis.getValue(...)` in this case, I think.

Even more, `TransferInput<Nullness, NullnessStore> input` has `input.getThenStore()`, which seems to work in this case:

```
Preconditions.checkArgument(a != null && !a.equals(this));
a.toString();
```

As printing `input.getThenStore()` inside a modified `handler.onDataflowVisitMethodInvocation(...)` gives:

```
Found Preconditions.checkArgument check at node: Preconditions.checkArgument(((a != null) && (!a.equals(this))))
inputThenStore: {AccessPath{root=a, elements=[], mapGetArg=null}=Non-null}
inputElseStore: {AccessPath{root=a, elements=[], mapGetArg=null}=Nullable}
```

Which we can use to set `thenUpdates`

The problem is, this **only** works for the complex boolean expression case (due to the implicit control flow).

In these cases:

```
Preconditions.checkArgument(a != null, "a ought to be non-null");
Preconditions.checkArgument(!Strings.isNullOrEmpty(a));
```

We simply get:

```
inputThenStore: {AccessPath{root=a, elements=[], mapGetArg=null}=Nullable}
```

As no branching is taking place inside the argument to the call.

**Comment by lazaroclapp** [comment]
Actually, never mind, **the above explanation is wrong**. 

The GOOD news is: that `input.getThenStore()` works for all three cases as desired!

The BAD news is: what breaks that approach is the presence of an explanation string in the `Preconditions` call, since in:

```
Preconditions.checkArgument(a != null && !a.equals(this), "a ought to be non-null");
```

The "previous" thing to be evaluated is the string "a ought to be non-null" and not `a != null && !a.equals(this)`

**Comment by msridhar** [comment]
> The "previous" thing to be evaluated is the string "a ought to be non-null" and not `a != null && !a.equals(this)`

I don't understand this.  I would expect that the approach would be to query the store(s) for the first actual parameter passed to `checkArgument`.  So why does the "previous" thing to be evaluated come into play?

**Comment by lazaroclapp** [comment]
> Another option is, having the node, try to find its corresponding block, and then use a reference to the analysis to call [getInput(Block b)](https://github.com/typetools/checker-framework/blob/95dd4483eeba563aa6755c9a4559a90db2e606a6/dataflow/src/main/java/org/checkerframework/dataflow/analysis/Analysis.java#L116)

This also doesn't seem to work, since the whole thing is in the same block, when asking about the block corresponding to `a != null`, I get:

```
Found Preconditions.checkArgument check at node: Preconditions.checkArgument((a != null), "a ought to be non-null")
firstArgBlock: RegularBlock([Preconditions.checkArgument, a, null, (a != null), "a ought to be non-null"])
Successors: [ExceptionBlock(Preconditions.checkArgument((a != null), "a ought to be non-null"))]
```

So, getting the store either before or after that whole `RegularBlock([Preconditions.checkArgument, a, null, (a != null), "a ought to be non-null"])` is no use to us. What we need is the store right after the `(a != null)` _node_, and I think we only have access to it from `visitMethodInvocation` in the case where that expression is the last argument and it just happen to coincidentally be the "input" node to the `invoke`.

**Comment by msridhar** [comment]
We could think about adding an API to Checker dataflow and opening a PR; I think they would be open to it if we show the change is useful.

**Comment by msridhar** [comment]
This issue is fixed in NullAway 0.9.8! Let us know if you still see problems.
